### PR TITLE
fix: No more crash when app was killed while uploading asset

### PR DESCRIPTION
**Pull request by ghost**
**Quick description**
While the application was uploading an asset, if the user decided to end the application process (app killed), the application would crash once when relaunching it.
To use again the application correctly, the user had to relaunch the application, wait for this crash, and then relaunch it once again, in order to resume the uploading process.

**Issue**
This crash is due to a class member of ListLibraryItems: _mLocalTempURL_
When a background task has finished (successfully or not), an URLSession delegate is called.
This delegates informs when a task has finished: it uses _mLocalTempURL_ to create a JSON, and sends it to the JS part.
When the user has ended the process of the application, _mLocalTempURL_ is deallocated, and the current background task is canceled.
When the user tries to relaunch the application, the delegates tries to inform to the JS part that the previous task has been canceled.
The app crashes because an exception is thrown. The JS part can't create the JSON since _mLocalTempURL_ was deallocated and equals to nil.

**Fix**
Now if the delegate is called, the application checks if _mLocalTempURL_ is nil, and if yes, the application will not inform the JS code.

### CODE REVIEW

File: src/ios/ListLibraryItems.m [hunk 1]
```diff
@@ -320,16 +318,21 @@ - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didComp
         NSString * errorMessage;
         
         if (error) {
+            NSLog(@"--- STATUS NOK ---");
```
Comment (y-lohse) [inline src/ios/ListLibraryItems.m]: I think we can remove this now :)

File: src/ios/ListLibraryItems.m [hunk 2]
```diff
@@ -167,8 +167,8 @@ - (void)uploadItem:(CDVInvokedUrlCommand *)command {
 
                             [request setValue:[[[NSFileManager defaultManager] sizeOfItemAtURL:mLocalTempURL] stringValue] forHTTPHeaderField:@"Content-Length"];
                             [request setValue:[[NSFileManager defaultManager] md5OfItemAtURL:mLocalTempURL] forHTTPHeaderField:@"Content-MD5"];
-                            NSURLSessionUploadTask * task = [session uploadTaskWithRequest:request fromFile:mLocalTempURL];
-                            [task resume];
+                            currentTask = [session uploadTaskWithRequest:request fromFile:mLocalTempURL];
+                            [currentTask resume];
```
Comment (y-lohse) [inline src/ios/ListLibraryItems.m]: I don't think these changes should be in this PR?

File: src/ios/ListLibraryItems.m [hunk 3]
```diff
@@ -14,6 +14,8 @@ @interface ListLibraryItems () {
     NSURL * mLocalTempURL;
     NSDictionary * mReceivedData;
     NSURLSession * session;
+    int64_t * receivedData;
+    NSURLSessionTask * currentTask;
```
Comment (y-lohse) [inline src/ios/ListLibraryItems.m]: I don't think these changes are part of this PR .

File: src/ios/ListLibraryItems.m [hunk 4]
```diff
@@ -320,16 +318,21 @@ - (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didComp
...
         }
         else {
             errorMessage = [NSHTTPURLResponse localizedStringForStatusCode:status];
         }
         NSLog(@"Error: %@", error);
-        NSMutableDictionary * json = [NSMutableDictionary dictionaryWithObjects:@[[NSString stringWithFormat:@"%ld", status], errorMessage, [mLocalTempURL absoluteString], target]
-                                                                        forKeys:@[@"code", @"message", @"source", @"target"]];
-        [self returnUploadResult:NO payload:json command:mCommand];
+        if(mLocalTempURL == nil) {
+            NSLog(@"The mLocalTempURL is nil, the app has been killed during upload.");
```
Comment (y-lohse) [inline src/ios/ListLibraryItems.m]: Instead of having a log that explains what is going on, maybe we can make the code more explicit? Something like

```
bool taskCanceledDueToAppKill = mLocalTempURL == nil && [error code] == NSTaskCanceled]

if (!taskCanceledDueToAppKill) {
  NSMutableDictionary * json etc…
  [self returnUploadResult:NO payload:json command:mCommand];
}
```

(pseudo code, i didn't test it). What do you think?

Comment (ghost) [reply src/ios/ListLibraryItems.m]: Understood, I guess this is a great idea and using this method we can really know as you said if the task is canceled due to app kill.

File: src/ios/ListLibraryItems.m (hunk 2 above)
Comment (ghost) [reply src/ios/ListLibraryItems.m]: You're right, all changes concerning this deserves to be in the timeout PR.

File: src/ios/ListLibraryItems.m [hunk 5]
```diff
@@ -102,6 +100,16 @@ - (void)returnUploadResult:(BOOL)aSuccess payload:(NSDictionary *)aJSON command:
...
 }
 
+- (bool)isTaskCanceledDueToAppKill:(NSURL*)aLocalTempURL error:(NSError *)anError status:(long)aStatus errormessage:(NSString *)anErrorMessage target:(NSString *)aTarget {
+    if(aLocalTempURL == nil && [anError code] == NSURLErrorCancelled) {
+        return true;
+    } else {
+        NSMutableDictionary * json = [NSMutableDictionary dictionaryWithObjects:@[[NSString stringWithFormat:@"%ld",aStatus], anErrorMessage,[aLocalTempURL absoluteString], aTarget]
+                                                                        forKeys:@[@"code", @"message", @"source", @"target"]];
+        [self returnUploadResult:NO payload:json command:mCommand];
+        return false;
```
Comment (Crash--) [inline src/ios/ListLibraryItems.m:107]: The else part should not be in that method. 

This method should only be responsible of the cancellation detection. Also, it should only receive `aLocalTempURL` and an `error`. No need to have the `status` or `message` since we can get it from the `error` object, right?

File: src/ios/ListLibraryItems.m [hunk 6]
```diff
@@ -167,7 +175,7 @@ - (void)uploadItem:(CDVInvokedUrlCommand *)command {
 
                             [request setValue:[[[NSFileManager defaultManager] sizeOfItemAtURL:mLocalTempURL] stringValue] forHTTPHeaderField:@"Content-Length"];
                             [request setValue:[[NSFileManager defaultManager] md5OfItemAtURL:mLocalTempURL] forHTTPHeaderField:@"Content-MD5"];
-                            NSURLSessionUploadTask * task = [session uploadTaskWithRequest:request fromFile:mLocalTempURL];
+                            NSURLSessionTask *task = [session uploadTaskWithRequest:request fromFile:mLocalTempURL];
```
Comment (Crash--) [inline src/ios/ListLibraryItems.m]: Why this change?

Comment (ghost) [reply src/ios/ListLibraryItems.m]: It was a typo, my bad. I just fixed it in the last commit.

File: src/ios/ListLibraryItems.m (hunk 5 above)
Comment (ghost) [reply src/ios/ListLibraryItems.m:107]: Fixed, I just left the two `return` in `if()` block. Now the JSON creation part is back in the delegate function !

**Review by Crash-- — APPROVED** [review]
Looks good to me! 

Let's merge it :)
